### PR TITLE
Fix/nex 1224/itempack set asset

### PR DESCRIPTION
**Pull request by siwane**
Add abililty to pack asset item asset one by one

### CODE REVIEW

File: models/classes/pack/ItemPack.php [hunk 1]
```diff
@@ -128,40 +130,50 @@ public function getData()
...
      * @param string $type the assets type, one of those who are supported.
-     * @param string[] $assets the list of assets' URL to load
-     *
-     * @param \tao_models_classes_service_StorageDirectory $publicDirectory
+     * @param array $assets the list of assets' URL to load
      *
      * @throw InvalidArgumentException
      */
-    public function setAssets($type, $assets, $publicDirectory)
+    public function setAssets(string $type, $assets): void
```
Comment (bartlomiejmarszal) [inline models/classes/pack/ItemPack.php:137]: This method is used by: [QtiItemPacker](https://github.com/oat-sa/extension-tao-itemqti/blob/0612d5a9617a0fdfeea6d48ed55f0d7924ab7c3b/model/pack/QtiItemPacker.php)

Comment (bartlomiejmarszal) [inline models/classes/pack/ItemPack.php:137]: This method is used by [QtiItemPacker](https://github.com/oat-sa/extension-tao-itemqti/blob/0612d5a9617a0fdfeea6d48ed55f0d7924ab7c3b/model/pack/QtiItemPacker.php)

File: models/classes/pack/ItemPack.php [hunk 2]
```diff
@@ -128,40 +130,50 @@ public function getData()
...
+                $assetKey = $asset->getMediaIdentifier();
             } else {
-                $assetKey = $asset;
+                $assetKey = $mediaSource->getBaseName($asset->getMediaIdentifier());
             }
 
-            $this->assets[$type][$assetKey] = $encoder->encode($asset);
+        } else {
+            $assetKey = $asset;
         }
```
Comment (bartlomiejmarszal) [inline models/classes/pack/ItemPack.php:174]: Can we extract this to a separate method?

File: models/classes/pack/ItemPack.php [hunk 3]
```diff
@@ -128,40 +130,50 @@ public function getData()
...
+    public function setAsset(string $type, $asset): void
+    {
+        if (!in_array($type, self::$assetTypes)) {
+            throw new InvalidArgumentException('Unknow asset type "' . $type . '", it should be either ' . implode(', ', self::$assetTypes));
+        }
+
+        /** @var Encoding $encoder */
+        $encoder = EncoderService::singleton()->get($this->assetEncoders[$type]);
+
+        if ($asset instanceof MediaAsset) {
```
Comment (gabrielfs7) [inline models/classes/pack/ItemPack.php:163]: I agree with @bartlomiejmarszal. We could make a method just to discover assetKey. This we we reduce method complexity (nested if/else) and is is easier to read. A suggestion:

```
    private function getAssetKey($asset): string
    {
        if (!$asset instanceof MediaAsset) {
            return $asset;
        }

        $mediaSource = $asset->getMediaSource();

        if ($mediaSource instanceof MediaSource || $mediaSource instanceof HttpSource) {
            return $asset->getMediaIdentifier();
        }

        return $mediaSource->getBaseName($asset->getMediaIdentifier());
    }
```

File: models/classes/pack/ItemPack.php (hunk 2 above)
Comment (gabrielfs7) [reply models/classes/pack/ItemPack.php:174]: I agree. Gave a suggestion: https://github.com/oat-sa/extension-tao-item/pull/410/files#r448169600

File: models/classes/pack/ItemPack.php (hunk 1 above)
Comment (siwane) [reply models/classes/pack/ItemPack.php:137]: it is not used here, having an extra parameters does not throw fatal error.
It is not a good practice it is why it is removed in https://github.com/oat-sa/extension-tao-itemqti/pull/1481

Comment (siwane) [reply models/classes/pack/ItemPack.php:137]: it is not used here. Having an extra parameters does not throw fatal error and the method can be call with this removed `$publicDirectory`
It is not a good practice it is why it is removed in https://github.com/oat-sa/extension-tao-itemqti/pull/1481

File: models/classes/pack/ItemPack.php (hunk 3 above)
Comment (siwane) [reply models/classes/pack/ItemPack.php]: no prob, it is good suggestion

**Comment by boajer** [comment]
- [ ] New code is covered by tests (if applicable)
- [ ] Tests are running successfully (old and new ones) on my local machine (if applicable)
- [ ] New code is respecting code style rules
- [ ] New code is respecting best practices
- [ ] New code is not subject to concurrency issues (if applicable)
- [ ] Feature is working correctly on my local machine (if applicable)
- [ ] Acceptance criteria are respected
- [ ] Pull request title and description are meaningful

**Comment by mike-ionut-mihai-sandu-tao** [comment]
- [ ] New code is covered by tests (if applicable)
- [ ] Tests are running successfully (old and new ones) on my local machine (if applicable)
- [ ] New code is respecting code style rules
- [ ] New code is respecting best practices
- [ ] New code is not subject to concurrency issues (if applicable)
- [ ] Feature is working correctly on my local machine (if applicable)
- [ ] Acceptance criteria are respected
- [ ] Pull request title and description are meaningful

File: models/classes/pack/ItemPack.php [hunk 4]
```diff
@@ -227,4 +228,19 @@ public function setNestedResourcesInclusion($nestedResourcesInclusion)
     {
         $this->nestedResourcesInclusion = (bool)$nestedResourcesInclusion;
     }
+
+    private function getAssetKey($asset): string
+    {
+        if (!$asset instanceof MediaAsset) {
+            return $asset;
```
[the remark this answers is not in the document]
Comment (siwane) [reply models/classes/pack/ItemPack.php:235]: It is the previous behaviour, but I can check what I can do. Thanks for the input

[the remark this answers is not in the document]
Comment (gabrielfs7) [reply models/classes/pack/ItemPack.php:246]: The validation should not be done here, though. But in the method you receive assets as input, cause there will still accept other types of objects.

[the remark this answers is not in the document]
Comment (siwane) [reply models/classes/pack/ItemPack.php:246]: As it will array key I prefer bart point of view

[the remark this answers is not in the document]
Comment (gabrielfs7) [reply models/classes/pack/ItemPack.php:246]: So you prefer that `setAsset` and `setAssets` do extra process even if $asset is not a supported value?

**Pull request merged**